### ISSUE TITLE
Duplication of entries in Source Files field when applying changes to yarnproject

### ISSUE THREAD

**Issue by H3G0A** [YarnSpinnerTool/YarnSpinner-Unity#293]
Whenever I make **any** change to a .yarnproject file in the Unity inspector, new entries appear in the "**Source Files**" field. The entries are always the same: **/*.yarn, and each time I repeat the process, more entries appear at once. 
If I create a fresh yarnproject it will start fine, with only 1 entry that says **/*.yarn and the first time I apply a change, 2 more of that same entry will apear. The next time, 6 will appear and each time that number doubles. This happens even if I delete all entries before applying changes. 
Furthermore, nothing happens when I click the button to add new entries myself (although it lets me "apply changes").

This behaviour only happens on my computer, but it happens no mater what Unity project I'm on. It doesn't even stop even after reinstalling unity/unity hub from scratch and cleaning any leftover files.

This problem began happening after some time using yarn spinner, it didn't occur at first. However, I don't recall commiting any significant changes related to yarn spinner during that time.

  
  - Yarn Spinner Version: 2.4.2
  - Unity Version: 2022.3.30f1

**Other information** 

This is the output from the file **Editor.log** each time a single change is applied:
`Asset Pipeline Refresh (id=751ed96754d0d2e4d8b6d05ab26719c5): Total: 0.008 seconds - Initiated by RefreshV2(AllowForceSynchronousImport)
Start importing Assets/8. Sandbox/Yarn Spinner/SandboxProject.yarnproject using Guid(bc1e161d314ee7a448d0e7937f80c1eb) Importer(-1,00000000000000000000000000000000)  -> (artifact id: '831a9f67d569009029f41c60154fe3f2') in 0.027275 seconds
Refreshing native plugins compatible for Editor in 19.85 ms, found 4 plugins.
Preloading 0 native plugins for Editor in 0.00 ms.
Start importing Assets/8. Sandbox/Yarn Spinner/SandboxProject.yarnproject using Guid(bc1e161d314ee7a448d0e7937f80c1eb) Importer(-1,00000000000000000000000000000000)  -> (artifact id: '831a9f67d569009029f41c60154fe3f2') in 0.025773 seconds
Refreshing native plugins compatible for Editor in 19.25 ms, found 4 plugins.
Preloading 0 native plugins for Editor in 0.00 ms.
Asset Pipeline Refresh (id=d27081a3f0edc134a982f6f813a6f4a0): Total: 0.198 seconds - Initiated by StopAssetImportingV2(NoUpdateAssetOptions)
[Worker0] Start importing Assets/8. Sandbox/Yarn Spinner/SandboxProject.yarnproject using Guid(bc1e161d314ee7a448d0e7937f80c1eb) Importer(815301076,1909f56bfc062723c751e8b465ee728b) 
[Worker0]  -> (artifact id: 'a1be9309e99c8b65c024234e7699884d') in 0.001889 seconds`

**Comment by desplesda** [comment]
Hi there,

We aren't able to reproduce this issue on our end. Can you please give us a bit more information about your setup?

- Which operating system are you using?
- If you're able, can you please paste the contents of your `.yarnproject` file?

Thanks!

**Comment by H3G0A** [comment]
- I'm using Windows 10.
- The issue happens with every new .yarnproject file that I create, no matter which one. Here is the content of the most recent one:

{
  "projectFileVersion": 2,
  "sourceFiles": [
    "**/*.yarn"
  ],
  "excludeFiles": [
    "**/*~/*"
  ],
  "localisation": {
    "es": {}
  },
  "baseLanguage": "en",
  "compilerOptions": {}
}

**Comment by TheMartianLife** [comment]
I'm going to give another attempt to reproducing this. I have a few questions, but if you'd like to get to the bottom of this I will probably have more and maybe the easiest way to chat about this would be [our Discord](https://discord.com/invite/yarnspinner).

* What are the permissions of the directory the project is in?
* Are the duplicate entries showing up in the Unity inspector, in the project file when examined as raw text, or both?

**Comment by TheMartianLife** [comment]
My first guess is that the issue could a combination of permissions issues and a logic error - where the logic error has never come up because it would only make a difference in cases where a project file or object could be written but not deleted.

[This line 156](https://github.com/YarnSpinnerTool/YarnSpinner-Unity/blob/abe45d82302aea7796d77e70296e89412ee884c2/Editor/Importers/YarnProjectImporter.cs#L156) here in `YarnProjectImporter.cs` means that when a new project object is created, instead of assigning the source file paths to its property it adds them to the existing value. The existing value should always be empty, but maybe in your case it's somehow getting a recycled version of the project or a prior value ??

```csharp
importData.sourceFilePaths.AddRange(project.SourceFilePatterns);
```

If you change that line to be...

```csharp
importData.sourceFilePaths = project.SourceFilePatterns;
```

And then remove all the duplicates from your yarn project object in editor and/or `.yarnproject` file it references on disk.

Do the duplicates return?

**Comment by TheMartianLife** [comment]
Weird, mine was totally fine with inferring that it should be a `List<string>`.

We actually pushed this change to the main branch by changing it to 

```csharp
importData.sourceFilePaths = new();
importData.sourceFilePaths.AddRange(project.SourceFilePatterns);
```

To ensure the list is initially empty.

If that doesn't work, maybe let me know if there's a way we could debug this more synchronously? I'd love to figure this out, and I have a feeling it's a weird one.

**Comment by H3G0A** [comment]
Yeah, I'll try with the commited change. Do I need to download a new release?

If this doesn't work either we cant talk through Discord as you suggested.

**Comment by TheMartianLife** [comment]
This change is just for the version you’re using. The newer versions of YS have actually renamed and changed some stuff in this importer.

But yeah, if you do this and delete all the duplicates and find they _still_ come back, then Discord would be great 🙂 [https://discord.com/invite/yarnspinner](https://discord.com/invite/yarnspinner)

**Comment by H3G0A** [comment]
I'm sorry, could you tell me how do I get said change into the yarn spinner version that's in my Unity project?
